File: EmployeeCostCalculator/ClientApp/src/app/services/discount-service.ts
import { Injectable } from '@angular/core';
import { DiscountRule } from '../models/discount-rule';

@Injectable({
  providedIn: 'root'
})
export class DiscountService {
  private readonly discountRules: DiscountRule[];

  constructor() {
    this.discountRules = [
      new DiscountRule(
        name => name.toLocaleLowerCase().startsWith('a'), 0.1)
    ];
  }

  //TODO: better use Person so can apply other rules
  public applyDiscount(name: string, cost: number): number {
    let result = cost;
    this.discountRules.forEach(r => {
      result = r.applyDiscount(name, result);
    });
    return result;
  }
}

File: CostServices/DiscountService.cs
using Entities;
using Interfaces;
using System;
using System.Collections.Generic;
using System.Text;

namespace CostServices
{
    public class DiscountService : IDiscountService
    {
        private readonly DiscountRule[] discountRules;

        public DiscountService()
        {
            this.discountRules = new[]{
                new DiscountRule(name => name.StartsWith("A", StringComparison.OrdinalIgnoreCase),
                0.1m)
            };
        }

        public decimal ApplyDiscount(string name, decimal cost)
        {
            var result = cost;
            foreach (var r in this.discountRules)
            {
                result = r.applyDiscount(name, result);
            }

            return result;
        }
    }
}
